libminifi/src/c2/ControllerSocketProtocol.cpp

```diff
@@ -66,7 +69,49 @@ ControllerSocketProtocol::ControllerSocketProtocol(core::controller::ControllerS
         configuration_(std::move(configuration)),
         socket_restart_processor_(update_sink_) {
   gsl_Expects(configuration_);
-  stream_factory_ = minifi::io::StreamFactory::getInstance(configuration_);
+}
+
+ControllerSocketProtocol::~ControllerSocketProtocol() {
+  stopListener();
+}
+
+void ControllerSocketProtocol::stopListener() {
+  io_context_.stop();
+  if (acceptor_) {
+    acceptor_->close();
+  }
+  if (server_thread_.joinable()) {
+    server_thread_.join();
+  }
+  io_context_.restart();
+}
+
+void ControllerSocketProtocol::startAccept() {
+  acceptor_->async_accept([this](const asio::error_code& error, asio::ip::tcp::socket socket) {
```

Comment:
I think you should consider using the coroutines instead.
That would remove the need for callbacks and as far as I know the performance is much better aswell.

Reply:
Thanks! Rewritten in da7844db28b0ee3a782ca2c2d2e1ffe4dfbebc73